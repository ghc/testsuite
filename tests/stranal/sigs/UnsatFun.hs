module UnsatFun where

-- Here we test how a partially applied function (f x)
-- with a bottom result affects the strictness signature
-- when used strictly (g) and lazily (g')
--
-- In both cases, the parameter x should not be absent

f :: Int -> Int -> Int
f x y = error (show x)
{-# NOINLINE f #-}

h :: (Int -> Int) -> Int
h f = f 2
{-# NOINLINE h #-}

h2 :: Bool -> (Int -> Int) -> Int
h2 True  _ = 0
h2 False f = f 2
{-# NOINLINE h2 #-}

-- Should get a bottom result
g :: Int -> Int
g x = let f' = f x
      in h f'

g2 :: Int -> Int
g2 x = let f' = f x
       in h2 True f'
