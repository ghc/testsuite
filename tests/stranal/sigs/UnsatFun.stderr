
==================== Strictness signatures ====================
UnsatFun.f: <B,1*U(U)><B,A>b
UnsatFun.g: <B,1*U(U)>b
UnsatFun.g2: <L,1*U(U)>
UnsatFun.h: <C(S),1*C1(U(U))>
UnsatFun.h2: <S,1*U><L,1*C1(U(U))>


